ssc/libssc.c

```diff
@@ -984,111 +1428,144 @@ static PetscErrorCode PCApply_PATCH(PC pc, Vec x, Vec y)
+                                        patch->patch_dof_weights[i]); CHKERRQ(ierr);
+
+            ierr = PCPatch_ScatterLocal_Private(pc, i + pStart,
+                                                patch->patchY[i], patch->localY,
+                                                ADD_VALUES, SCATTER_REVERSE); CHKERRQ(ierr);
+
+            /* Get the matrix on the patch but with only global bcs applied.
+             * This matrix is then multiplied with the result from the previous solve
+             * to obtain by how much the residual changes. */
+
+            if(patch->multiplicative){
+                ierr = MatMult(multMat, patch->patchY[i], patch->patchX[i]); CHKERRQ(ierr);
+                ierr = VecScale(patch->patchX[i], -1.0); CHKERRQ(ierr);
+                ierr = PCPatch_ScatterLocal_Private(pc, i + pStart,
+                                                    patch->patchX[i], patch->localX,
+                                                    ADD_VALUES, SCATTER_REVERSE); CHKERRQ(ierr);
+            }
+
+            if (patch->multiplicative && !patch->save_operators) {
+                ierr = MatDestroy(&multMat); CHKERRQ(ierr);
+            }
         }
-        ierr = ISBlockRestoreIndices(patch->bcs[i], &bcNodes); CHKERRQ(ierr);
-        ierr = VecRestoreArray(patch->patchX[i], &patchX); CHKERRQ(ierr);
-        if (!patch->save_operators) {
-            Mat mat;
-            ierr = PCPatchCreateMatrix(pc, patch->patchX[i], patch->patchY[i], &mat); CHKERRQ(ierr);
-            /* Populate operator here. */
-            ierr = PCPatchComputeOperator(pc, mat, i); CHKERRQ(ierr);
-            ierr = KSPSetOperators(patch->ksp[i], mat, mat);
-            /* Drop reference so the KSPSetOperators below will blow it away. */
-            ierr = MatDestroy(&mat); CHKERRQ(ierr);
-        }
-        ierr = PetscLogEventBegin(PC_Patch_Solve, pc, 0, 0, 0); CHKERRQ(ierr);
-        ierr = KSPSolve(patch->ksp[i], patch->patchX[i], patch->patchY[i]); CHKERRQ(ierr);
-        ierr = PetscLogEventEnd(PC_Patch_Solve, pc, 0, 0, 0); CHKERRQ(ierr);
-        if (!patch->save_operators) {
-            PC pc;
-            ierr = KSPSetOperators(patch->ksp[i], NULL, NULL); CHKERRQ(ierr);
-            ierr = KSPGetPC(patch->ksp[i], &pc); CHKERRQ(ierr);
-            /* Destroy PC context too, otherwise the factored matrix hangs around. */
-            ierr = PCReset(pc); CHKERRQ(ierr);
-        }
+    }
 
-        /* XXX: This bit needs changed for multiplicative combinations. */
-        /* XXX: pef thinks "do we not need to weight these
-         * contributions by the dof multiplicity?" */
-        ierr = PCPatch_ScatterLocal_Private(pc, i + pStart,
-                                            patch->patchY[i], patch->localY,
-                                            ADD_VALUES, SCATTER_REVERSE); CHKERRQ(ierr);
+    if (patch->partition_of_unity && !patch->multiplicative) {
+        /* XXX: should we do this on the global vector? */
```

Comment:
I don't think it matters.  Because we have the correct partition of unity on owned dofs (by definition)?